File: app/src/main/java/com/subrat/alarm/AlarmSoundService.java
package com.subrat.alarm;

import android.app.*;
import android.content.ContentResolver;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.media.AudioAttributes;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.IBinder;
import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;
import com.subrat.R;
import com.subrat.medicine.AddMedicineActivity;

import java.io.IOException;

/**
 * This is Background service class for the alarm and notification management after it get triggred by alarm that we set.
 */
public class AlarmSoundService extends Service {
    private MediaPlayer mediaPlayer;
    private NotificationManager alarmNotificationManager;
    private NotificationCompat.Builder mBuilder;

    //Notification ID for Alarm
    public static final int NOTIFICATION_ID = 1;
    public static final String NOTIFICATION_CHANNEL_ID = "10001";

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        playSound(this, getAlarmUri());
    }

    //play deafult alam sound
    private void playSound(Context context, Uri alert) {
        mediaPlayer = new MediaPlayer();
        try {
            mediaPlayer.setDataSource(context, alert);
            final AudioManager audioManager = (AudioManager) context
                    .getSystemService(Context.AUDIO_SERVICE);
            if (audioManager.getStreamVolume(AudioManager.STREAM_ALARM) != 0) {
                mediaPlayer.setAudioStreamType(AudioManager.STREAM_ALARM);
                mediaPlayer.prepare();
                mediaPlayer.start();
            }
        } catch (IOException e) {
            System.out.println("OOPS");
        }
    }

    private Uri getAlarmUri() {
        Uri alert = RingtoneManager
                .getDefaultUri(RingtoneManager.TYPE_ALARM);
        if (alert == null) {
            alert = RingtoneManager
                    .getDefaultUri(RingtoneManager.TYPE_NOTIFICATION);
            if (alert == null) {
                alert = RingtoneManager
                        .getDefaultUri(RingtoneManager.TYPE_RINGTONE);
            }
        }
        return alert;
    }


    @Override
    public void onDestroy() {
        super.onDestroy();

        //On destory stop and release the media player
        if (mediaPlayer != null && mediaPlayer.isPlaying()) {
            mediaPlayer.stop();
            mediaPlayer.reset();
            mediaPlayer.release();
        }
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        sendNotification(intent.getStringExtra("medicine"), intent.getLongExtra("id", -1), this);
        return super.onStartCommand(intent, flags, startId);
    }

    //handle notification
    private void sendNotification(String msg, Long id, Context context) {
        alarmNotificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        Uri sound = Uri.parse(ContentResolver.SCHEME_ANDROID_RESOURCE + "://" + context.getPackageName() + "/" + R.raw.alarm_sound);
        //get pending intent
        Intent intent = new Intent(context, AddMedicineActivity.class);
        intent.putExtra("id", id);
        PendingIntent contentIntent = PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);


        //Create notification
        mBuilder = new NotificationCompat.Builder(context)
                .setContentTitle("Medicine Time")
                .setSmallIcon(R.mipmap.ic_launcher)
                .setStyle(new NotificationCompat.BigTextStyle().bigText(msg))
                .setContentText(msg)
                .setAutoCancel(true)
                .setContentIntent(contentIntent);

        //notiy notification manager about new notification
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            AudioAttributes attributes = new AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_NOTIFICATION)
                    .build();
            int importance = NotificationManager.IMPORTANCE_DEFAULT;
            NotificationChannel notificationChannel = new NotificationChannel(NOTIFICATION_CHANNEL_ID, "NOTIFICATION_CHANNEL_NAME", importance);
            notificationChannel.enableLights(true);
            notificationChannel.setLightColor(Color.RED);
            notificationChannel.enableVibration(true);
            notificationChannel.setVibrationPattern(new long[]{100, 200, 300, 400, 500, 400, 300, 200, 400});
            assert alarmNotificationManager != null;
            mBuilder.setChannelId(NOTIFICATION_CHANNEL_ID);
            alarmNotificationManager.createNotificationChannel(notificationChannel);
        }
        assert alarmNotificationManager != null;
        Notification notification = mBuilder.build();
        alarmNotificationManager.notify(NOTIFICATION_ID, notification);
        startForeground(NOTIFICATION_ID, notification);
    }
}

File: app/src/main/java/com/subrat/alarm/AlarmSoundService.kt
package com.subrat.alarm

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.ContentResolver
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.media.AudioAttributes
import android.media.AudioManager
import android.media.MediaPlayer
import android.media.RingtoneManager
import android.net.Uri
import android.os.IBinder
import androidx.core.app.NotificationCompat
import com.subrat.R
import com.subrat.medicine.AddMedicineActivity
import java.io.IOException

/**
 * This is Background service class for the alarm and notification management after it get triggred by alarm that we set.
 */
class AlarmSoundService : Service() {
    private var mediaPlayer: MediaPlayer? = null
    private var alarmNotificationManager: NotificationManager? = null
    private var mBuilder: NotificationCompat.Builder? = null

    private val alarmUri: Uri?
        get() {
            var alert: Uri? = RingtoneManager
                .getDefaultUri(RingtoneManager.TYPE_ALARM)
            if (alert == null) {
                alert = RingtoneManager
                    .getDefaultUri(RingtoneManager.TYPE_NOTIFICATION)
                if (alert == null) {
                    alert = RingtoneManager
                        .getDefaultUri(RingtoneManager.TYPE_RINGTONE)
                }
            }
            return alert
        }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onCreate() {
        super.onCreate()
        playSound(this, alarmUri)
    }

    //play deafult alam sound
    private fun playSound(context: Context, alert: Uri?) {
        mediaPlayer = MediaPlayer()
        try {
            mediaPlayer!!.setDataSource(context, alert!!)
            val audioManager = context
                .getSystemService(Context.AUDIO_SERVICE) as AudioManager
            if (audioManager.getStreamVolume(AudioManager.STREAM_ALARM) != 0) {
                mediaPlayer!!.setAudioStreamType(AudioManager.STREAM_ALARM)
                mediaPlayer!!.prepare()
                mediaPlayer!!.start()
            }
        } catch (e: IOException) {
            println("OOPS")
        }

    }


    override fun onDestroy() {
        super.onDestroy()

        //On destory stop and release the media player
        if (mediaPlayer != null && mediaPlayer!!.isPlaying) {
            mediaPlayer!!.stop()
            mediaPlayer!!.reset()
            mediaPlayer!!.release()
        }
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        sendNotification(intent.getStringExtra("medicine"), intent.getLongExtra("id", -1), this)
        return super.onStartCommand(intent, flags, startId)
    }

    //handle notification
    private fun sendNotification(msg: String, id: Long?, context: Context) {
        alarmNotificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val sound =
            Uri.parse(ContentResolver.SCHEME_ANDROID_RESOURCE + "://" + context.packageName + "/" + R.raw.alarm_sound)
        //get pending intent
        val intent = Intent(context, AddMedicineActivity::class.java)
        intent.putExtra("id", id)
        val contentIntent = PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT)


        //Create notification
        mBuilder = NotificationCompat.Builder(context)
            .setContentTitle("Medicine Time")
            .setSmallIcon(R.mipmap.ic_launcher)
            .setStyle(NotificationCompat.BigTextStyle().bigText(msg))
            .setContentText(msg)
            .setAutoCancel(true)
            .setContentIntent(contentIntent)

        //notiy notification manager about new notification
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            val attributes = AudioAttributes.Builder()
                .setUsage(AudioAttributes.USAGE_NOTIFICATION)
                .build()
            val importance = NotificationManager.IMPORTANCE_DEFAULT
            val notificationChannel =
                NotificationChannel(NOTIFICATION_CHANNEL_ID, "NOTIFICATION_CHANNEL_NAME", importance)
            notificationChannel.enableLights(true)
            notificationChannel.lightColor = Color.RED
            notificationChannel.enableVibration(true)
            notificationChannel.vibrationPattern = longArrayOf(100, 200, 300, 400, 500, 400, 300, 200, 400)
            assert(alarmNotificationManager != null)
            mBuilder!!.setChannelId(NOTIFICATION_CHANNEL_ID)
            alarmNotificationManager!!.createNotificationChannel(notificationChannel)
        }
        assert(alarmNotificationManager != null)
        val notification = mBuilder!!.build()
        alarmNotificationManager!!.notify(NOTIFICATION_ID, notification)
        startForeground(NOTIFICATION_ID, notification)
    }

    companion object {

        //Notification ID for Alarm
        val NOTIFICATION_ID = 1
        val NOTIFICATION_CHANNEL_ID = "10001"
    }
}
